### ISSUE TITLE
Proposal: ParenthesizedType

### ISSUE THREAD

**Issue by k2snowman69** [estree/estree#193]
This proposal is to add a parenthesis type for expressions. This was slightly mentioned in the [CST issue](https://github.com/estree/estree/issues/41) but I think this is much more basic, hence splitting it out into it's own issue.

### Issue
Note: yes this is very contrived and will always result in the same value but it's trimmed down for this example:
```
let value = (2) | 1;
```

The trimmed down output of this without a parenthesized type is:
```
{
    type: "BinaryExpression",
    range: [12, 19],
    operator: "|",
    left: {
      type: "Literal",
      range: [13, 14],
      value: 2,
      raw: "2"
    },
    right: {
      type: "Literal",
      range: [18, 19],
      value: 1,
      raw: "1"
    }
  };
```
which means the context of the parenthesis is missing. This means if you were to reprint this document, you have no context as to where a user may or may not have added parenthesis to break up expressions.

### Proposal
Adding a ParenthesizedType will allow for the context of the parenthesis to be conveyed in the AST
```
{
    type: "BinaryExpression",
    range: [12, 19],
    operator: "|",
    left: {
      type: "ParenthesisExpression",
      range: [12, 15],
      raw: "(2)",
      value: {
        type: "Literal",
        range: [13, 14],
        value: 2,
        raw: "2"
      }
    },
    right: {
      type: "Literal",
      range: [18, 19],
      value: 1,
      raw: "1"
    }
  }
```

Credit to where credit is due, I got this idea from TSParenthesizedType from [typescript-estree](https://github.com/JamesHenry/typescript-estree/pull/55)

**Comment by Meir017** [comment]
is this related to the following scenario?
```js
(fn1) = function(){}
fn2 = function(){}
```
![image](https://user-images.githubusercontent.com/9786571/167780894-052bcf29-a7bb-4199-976a-793d1c6fed91.png)
viewing this in [astexplorer.net](https://astexplorer.net/#/gist/8eadd529ddd1d363590d5ec4be58e089/1dc1230ada392bd140fe16b8ccb4058f80a29707) shows identical trees for both
![image](https://user-images.githubusercontent.com/9786571/167781066-72f69640-7ba0-4ea5-9b1a-de1578d09d8c.png)

**Comment by k2snowman69** [comment]
Yep, exactly. These two trees should not be identical and instead for the left identifier there should be a `ParenthesizedType`

**Comment by Meir017** [comment]
related babel issue recently fixed - https://github.com/babel/babel/issues/14514
